ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/CommandStatusFuture.java

```diff
@@ -19,24 +19,17 @@
 import io.confluent.ksql.rest.entity.CommandStatus;
 import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.TimeoutException;
-import java.util.concurrent.atomic.AtomicReference;
-import org.apache.kafka.common.utils.Utils;
-
 
 public class CommandStatusFuture implements Future<CommandStatus> {
-  public interface Cleanup {
-    void run(CommandId commandId);
-  }
-
   private final CommandId commandId;
-  private Cleanup cleanup;
-  private final AtomicReference<CommandStatus> result;
+  private boolean completed;
```

Comment:
Don't these need to be `volatile`? If they are you can remove the `synchronized` you've added to the  methods.

Can you explain your reasoning for moving away from `AtomicReference`, which is designed to be thread-safe, and moving to your own `completed` flag?

Splitting the completion into a setter and a completion method also seems a little strange - can you explain why this is needed?

Reply:
> Don't these need to be volatile? If they are you can remove the synchronized you've added to the methods.

Right but if we're synchronizing access then we don't need these to be volatile.

> Can you explain your reasoning for moving away from AtomicReference, which is designed to be thread-safe, and moving to your own completed flag?

To distinguish between non-terminal and terminal status changes.

> Splitting the completion into a setter and a completion method also seems a little strange - can you explain why this is needed?

See the above comment.

Reply:
Using `volatile` over `synchronized` methods can result in less blocking / locking, as its just a memory barrier and not a full on critical section. However, it's no biggie unless this is in critical code.

Having said that, the code was probably safer / easy to understand / maintain when using the `AtomicReference`. So if you go with refactoring as I mentioned above, i.e. adding a new object that handles both the intermediate state and holds a `CompletableFuture<CommandStatus>` for the final result, then I think this all becomes academic anyway ;)